### ISSUE TITLE
Invalid characters in OpenDAP URL

### ISSUE THREAD

**Issue by pmartineauGit** [pydata/xarray#5913]
Hello,

I have successfully opened an OpenDAP URL with ds = xarray.open_dataset(url)
However, after selecting a subset with ds = ds.isel(time=0) and attempting to
load the data with ds.load(), I get the following error: 

HTTP Status 400 – Bad Request: Invalid character found in the request
> target. The valid characters are defined in RFC 7230 and RFC 3986

I suspect the reason is that square brackets are passed in the URL when attempting to load:
...zg_6hrPlevPt_MIROC6_historical_r1i1p1f1_gn_185001010600-185101010000.nc.dods?zg.zg[0][0:6][0:127][0:255]]
because of the index selection with .isel()

In fact, some servers do forbid square brackets: https://www.unidata.ucar.edu/mailing_lists/archives/thredds/2020/msg00056.html

Would it be possible to provide an option to encode URLs? ( [ becomes %5B, and ] becomes %5D )

Or, instead of loading directly with ds.load(), is there a way for me to retrieve the URL with offending brackets that is generated automatically by xarray, encode it myself, and then use ds2 = xarray.load_dataset(encoded_url) to load?

Thank you for your help!

**Comment by dopplershift** [comment]
Is this with the netcdf or with the pydap engine? If you're not sure, can you post the full error traceback?

**Comment by pmartineauGit** [comment]
This was using the netcdf4 engine.

Here's the full error in case it is useful:

syntax error, unexpected WORD_WORD, expecting SCAN_ATTR or SCAN_DATASET or SCAN_ERROR
context: <!doctype^ html><html lang="en"><head><title>HTTP Status 400 – Bad Request</title><style type="text/css">body {font-family:Tahoma,Arial,sans-serif;} h1, h2, h3, b {color:white;background-color:#525D76;} h1 {font-size:22px;} h2 {font-size:16px;} h3 {font-size:14px;} p {font-size:12px;} a {color:black;} .line {height:1px;background-color:#525D76;border:none;}</style></head><body><h1>HTTP Status 400 – Bad Request</h1><hr class="line" /><p><b>Type</b> Exception Report</p><p><b>Message</b> Invalid character found in the request target [&#47;thredds&#47;dodsC&#47;css03_data&#47;CMIP6&#47;CMIP&#47;MIROC&#47;MIROC6&#47;historical&#47;r1i1p1f1&#47;6hrPlevPt&#47;zg&#47;gn&#47;v20191204&#47;zg_6hrPlevPt_MIROC6_historical_r1i1p1f1_gn_185001010600-185101010000.nc.dods?zg.zg[359:478][0:6][0:127][0:255]]. The valid characters are defined in RFC 7230 and RFC 3986</p><p><b>Description</b> The server cannot or will not process the request due to something that is perceived to be a client error (e.g., malformed request syntax, invalid request message framing, or deceptive request routing).</p><p><b>Exception</b></p><pre>java.lang.IllegalArgumentException: Invalid character found in the request target [&#47;thredds&#47;dodsC&#47;css03_data&#47;CMIP6&#47;CMIP&#47;MIROC&#47;MIROC6&#47;historical&#47;r1i1p1f1&#47;6hrPlevPt&#47;zg&#47;gn&#47;v20191204&#47;zg_6hrPlevPt_MIROC6_historical_r1i1p1f1_gn_185001010600-185101010000.nc.dods?zg.zg[359:478][0:6][0:127][0:255]]. The valid characters are defined in RFC 7230 and RFC 3986 org.apache.coyote.http11.Http11InputBuffer.parseRequestLine(Http11InputBuffer.java:505) org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:502) org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:65) org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:818) org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1626) org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:49) java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61) java.lang.Thread.run(Thread.java:748)</pre><p><b>Note</b> The full stack trace of the root cause is available in the server logs.</p><hr class="line" /><h3>Apache Tomcat/8.5.57</h3></body></html>
syntax error, unexpected WORD_WORD, expecting SCAN_ATTR or SCAN_DATASET or SCAN_ERROR
context: <!doctype^ html><html lang="en"><head><title>HTTP Status 400 – Bad Request</title><style type="text/css">h1 {font-family:Tahoma,Arial,sans-serif;color:white;background-color:#525D76;font-size:22px;} h2 {font-family:Tahoma,Arial,sans-serif;color:white;background-color:#525D76;font-size:16px;} h3 {font-family:Tahoma,Arial,sans-serif;color:white;background-color:#525D76;font-size:14px;} body {font-family:Tahoma,Arial,sans-serif;color:black;background-color:white;} b {font-family:Tahoma,Arial,sans-serif;color:white;background-color:#525D76;} p {font-family:Tahoma,Arial,sans-serif;background:white;color:black;font-size:12px;} a {color:black;} a.name {color:black;} .line {height:1px;background-color:#525D76;border:none;}</style></head><body><h1>HTTP Status 400 – Bad Request</h1><hr class="line" /><p><b>Type</b> Exception Report</p><p><b>Message</b> Invalid character found in the request target. The valid characters are defined in RFC 7230 and RFC 3986</p><p><b>Description</b> The server cannot or will not process the request due to something that is perceived to be a client error (e.g., malformed request syntax, invalid request message framing, or deceptive request routing).</p><p><b>Exception</b></p><pre>java.lang.IllegalArgumentException: Invalid character found in the request target. The valid characters are defined in RFC 7230 and RFC 3986 org.apache.coyote.http11.Http11InputBuffer.parseRequestLine(Http11InputBuffer.java:483) org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:682) org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:66) org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:810) org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1500) org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:49) java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61) java.lang.Thread.run(Thread.java:748)</pre><p><b>Note</b> The full stack trace of the root cause is available in the server logs.</p><hr class="line" /><h3>Apache Tomcat/8.5.46</h3></body></html>
syntax error, unexpected WORD_WORD, expecting SCAN_ATTR or SCAN_DATASET or SCAN_ERROR
context: <!doctype^ html><html lang="en"><head><title>HTTP Status 400 – Bad Request</title><style type="text/css">h1 {font-family:Tahoma,Arial,sans-serif;color:white;background-color:#525D76;font-size:22px;} h2 {font-family:Tahoma,Arial,sans-serif;color:white;background-color:#525D76;font-size:16px;} h3 {font-family:Tahoma,Arial,sans-serif;color:white;background-color:#525D76;font-size:14px;} body {font-family:Tahoma,Arial,sans-serif;color:black;background-color:white;} b {font-family:Tahoma,Arial,sans-serif;color:white;background-color:#525D76;} p {font-family:Tahoma,Arial,sans-serif;background:white;color:black;font-size:12px;} a {color:black;} a.name {color:black;} .line {height:1px;background-color:#525D76;border:none;}</style></head><body><h1>HTTP Status 400 – Bad Request</h1><hr class="line" /><p><b>Type</b> Exception Report</p><p><b>Message</b> Invalid character found in the request target. The valid characters are defined in RFC 7230 and RFC 3986</p><p><b>Description</b> The server cannot or will not process the request due to something that is perceived to be a client error (e.g., malformed request syntax, invalid request message framing, or deceptive request routing).</p><p><b>Exception</b></p><pre>java.lang.IllegalArgumentException: Invalid character found in the request target. The valid characters are defined in RFC 7230 and RFC 3986 org.apache.coyote.http11.Http11InputBuffer.parseRequestLine(Http11InputBuffer.java:479) org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:684) org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:66) org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:806) org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1498) org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:49) java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61) java.lang.Thread.run(Thread.java:748)</pre><p><b>Note</b> The full stack trace of the root cause is available in the server logs.</p><hr class="line" /><h3>Apache Tomcat/8.5.39</h3></body></html>
syntax error, unexpected WORD_WORD, expecting SCAN_ATTR or SCAN_DATASET or SCAN_ERROR
context: <!doctype^ html><html lang="en"><head><title>HTTP Status 400 – Bad Request</title><style type="text/css">body {font-family:Tahoma,Arial,sans-serif;} h1, h2, h3, b {color:white;background-color:#525D76;} h1 {font-size:22px;} h2 {font-size:16px;} h3 {font-size:14px;} p {font-size:12px;} a {color:black;} .line {height:1px;background-color:#525D76;border:none;}</style></head><body><h1>HTTP Status 400 – Bad Request</h1><hr class="line" /><p><b>Type</b> Exception Report</p><p><b>Message</b> Invalid character found in the request target [&#47;thredds&#47;dodsC&#47;css03_data&#47;CMIP6&#47;CMIP&#47;MIROC&#47;MIROC6&#47;historical&#47;r1i1p1f1&#47;6hrPlevPt&#47;zg&#47;gn&#47;v20191204&#47;zg_6hrPlevPt_MIROC6_historical_r1i1p1f1_gn_185001010600-185101010000.nc.dods?zg.zg[359:478][0:6][0:127][0:255]]. The valid characters are defined in RFC 7230 and RFC 3986</p><p><b>Description</b> The server cannot or will not process the request due to something that is perceived to be a client error (e.g., malformed request syntax, invalid request message framing, or deceptive request routing).</p><p><b>Exception</b></p><pre>java.lang.IllegalArgumentException: Invalid character found in the request target [&#47;thredds&#47;dodsC&#47;css03_data&#47;CMIP6&#47;CMIP&#47;MIROC&#47;MIROC6&#47;historical&#47;r1i1p1f1&#47;6hrPlevPt&#47;zg&#47;gn&#47;v20191204&#47;zg_6hrPlevPt_MIROC6_historical_r1i1p1f1_gn_185001010600-185101010000.nc.dods?zg.zg[359:478][0:6][0:127][0:255]]. The valid characters are defined in RFC 7230 and RFC 3986 org.apache.coyote.http11.Http11InputBuffer.parseRequestLine(Http11InputBuffer.java:505) org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:502) org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:65) org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:818) org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1626) org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:49) java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61) java.lang.Thread.run(Thread.java:748)</pre><p><b>Note</b> The full stack trace of the root cause is available in the server logs.</p><hr class="line" /><h3>Apache Tomcat/8.5.57</h3></body></html>
syntax error, unexpected WORD_WORD, expecting SCAN_ATTR or SCAN_DATASET or SCAN_ERROR
context: <!doctype^ html><html lang="en"><head><title>HTTP Status 400 – Bad Request</title><style type="text/css">h1 {font-family:Tahoma,Arial,sans-serif;color:white;background-color:#525D76;font-size:22px;} h2 {font-family:Tahoma,Arial,sans-serif;color:white;background-color:#525D76;font-size:16px;} h3 {font-family:Tahoma,Arial,sans-serif;color:white;background-color:#525D76;font-size:14px;} body {font-family:Tahoma,Arial,sans-serif;color:black;background-color:white;} b {font-family:Tahoma,Arial,sans-serif;color:white;background-color:#525D76;} p {font-family:Tahoma,Arial,sans-serif;background:white;color:black;font-size:12px;} a {color:black;} a.name {color:black;} .line {height:1px;background-color:#525D76;border:none;}</style></head><body><h1>HTTP Status 400 – Bad Request</h1><hr class="line" /><p><b>Type</b> Exception Report</p><p><b>Message</b> Invalid character found in the request target. The valid characters are defined in RFC 7230 and RFC 3986</p><p><b>Description</b> The server cannot or will not process the request due to something that is perceived to be a client error (e.g., malformed request syntax, invalid request message framing, or deceptive request routing).</p><p><b>Exception</b></p><pre>java.lang.IllegalArgumentException: Invalid character found in the request target. The valid characters are defined in RFC 7230 and RFC 3986 org.apache.coyote.http11.Http11InputBuffer.parseRequestLine(Http11InputBuffer.java:483) org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:682) org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:66) org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:810) org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1500) org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:49) java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61) java.lang.Thread.run(Thread.java:748)</pre><p><b>Note</b> The full stack trace of the root cause is available in the server logs.</p><hr class="line" /><h3>Apache Tomcat/8.5.46</h3></body></html>
syntax error, unexpected WORD_WORD, expecting SCAN_ATTR or SCAN_DATASET or SCAN_ERROR
context: <!doctype^ html><html lang="en"><head><title>HTTP Status 400 – Bad Request</title><style type="text/css">body {font-family:Tahoma,Arial,sans-serif;} h1, h2, h3, b {color:white;background-color:#525D76;} h1 {font-size:22px;} h2 {font-size:16px;} h3 {font-size:14px;} p {font-size:12px;} a {color:black;} .line {height:1px;background-color:#525D76;border:none;}</style></head><body><h1>HTTP Status 400 – Bad Request</h1><hr class="line" /><p><b>Type</b> Exception Report</p><p><b>Message</b> Invalid character found in the request target [&#47;thredds&#47;dodsC&#47;css03_data&#47;CMIP6&#47;CMIP&#47;MIROC&#47;MIROC6&#47;historical&#47;r1i1p1f1&#47;6hrPlevPt&#47;zg&#47;gn&#47;v20191204&#47;zg_6hrPlevPt_MIROC6_historical_r1i1p1f1_gn_185001010600-185101010000.nc.dods?zg.zg[359:478][0:6][0:127][0:255]]. The valid characters are defined in RFC 7230 and RFC 3986</p><p><b>Description</b> The server cannot or will not process the request due to something that is perceived to be a client error (e.g., malformed request syntax, invalid request message framing, or deceptive request routing).</p><p><b>Exception</b></p><pre>java.lang.IllegalArgumentException: Invalid character found in the request target [&#47;thredds&#47;dodsC&#47;css03_data&#47;CMIP6&#47;CMIP&#47;MIROC&#47;MIROC6&#47;historical&#47;r1i1p1f1&#47;6hrPlevPt&#47;zg&#47;gn&#47;v20191204&#47;zg_6hrPlevPt_MIROC6_historical_r1i1p1f1_gn_185001010600-185101010000.nc.dods?zg.zg[359:478][0:6][0:127][0:255]]. The valid characters are defined in RFC 7230 and RFC 3986 org.apache.coyote.http11.Http11InputBuffer.parseRequestLine(Http11InputBuffer.java:505) org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:502) org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:65) org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:818) org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1626) org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:49) java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61) java.lang.Thread.run(Thread.java:748)</pre><p><b>Note</b> The full stack trace of the root cause is available in the server logs.</p><hr class="line" /><h3>Apache Tomcat/8.5.57</h3></body></html>
syntax error, unexpected WORD_WORD, expecting SCAN_ATTR or SCAN_DATASET or SCAN_ERROR
context: <!doctype^ html><html lang="en"><head><title>HTTP Status 400 – Bad Request</title><style type="text/css">h1 {font-family:Tahoma,Arial,sans-serif;color:white;background-color:#525D76;font-size:22px;} h2 {font-family:Tahoma,Arial,sans-serif;color:white;background-color:#525D76;font-size:16px;} h3 {font-family:Tahoma,Arial,sans-serif;color:white;background-color:#525D76;font-size:14px;} body {font-family:Tahoma,Arial,sans-serif;color:black;background-color:white;} b {font-family:Tahoma,Arial,sans-serif;color:white;background-color:#525D76;} p {font-family:Tahoma,Arial,sans-serif;background:white;color:black;font-size:12px;} a {color:black;} a.name {color:black;} .line {height:1px;background-color:#525D76;border:none;}</style></head><body><h1>HTTP Status 400 – Bad Request</h1><hr class="line" /><p><b>Type</b> Exception Report</p><p><b>Message</b> Invalid character found in the request target. The valid characters are defined in RFC 7230 and RFC 3986</p><p><b>Description</b> The server cannot or will not process the request due to something that is perceived to be a client error (e.g., malformed request syntax, invalid request message framing, or deceptive request routing).</p><p><b>Exception</b></p><pre>java.lang.IllegalArgumentException: Invalid character found in the request target. The valid characters are defined in RFC 7230 and RFC 3986 org.apache.coyote.http11.Http11InputBuffer.parseRequestLine(Http11InputBuffer.java:483) org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:682) org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:66) org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:810) org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1500) org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:49) java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61) java.lang.Thread.run(Thread.java:748)</pre><p><b>Note</b> The full stack trace of the root cause is available in the server logs.</p><hr class="line" /><h3>Apache Tomcat/8.5.46</h3></body></html>
---------------------------------------------------------------------------
RuntimeError                          	Traceback (most recent call last)
/mnt/kuroshio/atsumi/patrick/CMIP/CMIP_DLD2.py in <module>
	165
	166                 	print('creating delayed task')
--> 167                 	delayed = xr.save_mfdataset(dsl, ptl, compute=False)
	168                 	print('computing')
	169                 	delayed.compute()

~/anaconda3/envs/my37/lib/python3.7/site-packages/xarray/backends/api.py in save_mfdataset(datasets, paths, mode, format, groups, engine, compute)
   1228             	ds, path, mode, format, group, engine, compute=compute, multifile=True
   1229         	)
-> 1230         	for ds, path, group in zip(datasets, paths, groups)
   1231     	]
   1232 	)

~/anaconda3/envs/my37/lib/python3.7/site-packages/xarray/backends/api.py in <listcomp>(.0)
   1228             	ds, path, mode, format, group, engine, compute=compute, multifile=True
   1229         	)
-> 1230         	for ds, path, group in zip(datasets, paths, groups)
   1231     	]
   1232 	)

~/anaconda3/envs/my37/lib/python3.7/site-packages/xarray/backends/api.py in to_netcdf(dataset, path_or_file, mode, format, group, engine, encoding, unlimited_dims, compute, multifile, invalid_netcdf)
   1076     	# to be parallelized with dask
   1077     	dump_to_store(
-> 1078         	dataset, store, writer, encoding=encoding, unlimited_dims=unlimited_dims
   1079     	)
   1080     	if autoclose:

~/anaconda3/envs/my37/lib/python3.7/site-packages/xarray/backends/api.py in dump_to_store(dataset, store, writer, encoder, encoding, unlimited_dims)
   1122     	variables, attrs = encoder(variables, attrs)
   1123
-> 1124 	store.store(variables, attrs, check_encoding, writer, unlimited_dims=unlimited_dims)
   1125
   1126

~/anaconda3/envs/my37/lib/python3.7/site-packages/xarray/backends/common.py in store(self, variables, attributes, check_encoding_set, writer, unlimited_dims)
	260         	writer = ArrayWriter()
	261
--> 262     	variables, attributes = self.encode(variables, attributes)
	263
	264     	self.set_attributes(attributes)

~/anaconda3/envs/my37/lib/python3.7/site-packages/xarray/backends/common.py in encode(self, variables, attributes)
	349     	# All NetCDF files get CF encoded by default, without this attempting
	350     	# to write times, for example, would fail.
--> 351     	variables, attributes = cf_encoder(variables, attributes)
	352     	variables = {k: self.encode_variable(v) for k, v in variables.items()}
	353     	attributes = {k: self.encode_attribute(v) for k, v in attributes.items()}

~/anaconda3/envs/my37/lib/python3.7/site-packages/xarray/conventions.py in cf_encoder(variables, attributes)
	851
	852 	# add encoding for time bounds variables if present.
--> 853 	_update_bounds_encoding(variables)
	854
	855 	new_vars = {k: encode_cf_variable(v, name=k) for k, v in variables.items()}

~/anaconda3/envs/my37/lib/python3.7/site-packages/xarray/conventions.py in _update_bounds_encoding(variables)
	431     	is_datetime_type = np.issubdtype(
	432         	v.dtype, np.datetime64
--> 433     	) or contains_cftime_datetimes(v)
	434
	435     	if (

~/anaconda3/envs/my37/lib/python3.7/site-packages/xarray/core/common.py in contains_cftime_datetimes(var)
   1839 def contains_cftime_datetimes(var) -> bool:
   1840 	"""Check if an xarray.Variable contains cftime.datetime objects"""
-> 1841 	return _contains_cftime_datetimes(var.data)
   1842
   1843

~/anaconda3/envs/my37/lib/python3.7/site-packages/xarray/core/variable.py in data(self)
	342         	return self._data
	343     	else:
--> 344         	return self.values
	345
	346 	@data.setter

~/anaconda3/envs/my37/lib/python3.7/site-packages/xarray/core/variable.py in values(self)
	515 	def values(self):
	516     	"""The variable's data as a numpy.ndarray"""
--> 517     	return _as_array_or_item(self._data)
	518
	519 	@values.setter

~/anaconda3/envs/my37/lib/python3.7/site-packages/xarray/core/variable.py in _as_array_or_item(data)
	257 	TODO: remove this (replace with np.asarray) once these issues are fixed
	258 	"""
--> 259 	data = np.asarray(data)
	260 	if data.ndim == 0:
	261     	if data.dtype.kind == "M":

~/anaconda3/envs/my37/lib/python3.7/site-packages/numpy/core/_asarray.py in asarray(a, dtype, order)
 	81
 	82 	"""
---> 83 	return array(a, dtype, copy=False, order=order)
 	84
 	85

~/anaconda3/envs/my37/lib/python3.7/site-packages/xarray/core/indexing.py in __array__(self, dtype)
	417 	def __array__(self, dtype=None):
	418     	array = as_indexable(self.array)
--> 419     	return np.asarray(array[self.key], dtype=None)
	420
	421 	def transpose(self, order):

~/anaconda3/envs/my37/lib/python3.7/site-packages/numpy/core/_asarray.py in asarray(a, dtype, order)
 	81
 	82 	"""
---> 83 	return array(a, dtype, copy=False, order=order)
 	84
 	85

~/anaconda3/envs/my37/lib/python3.7/site-packages/xarray/coding/variables.py in __array__(self, dtype)
 	68
 	69 	def __array__(self, dtype=None):
---> 70     	return self.func(self.array)
 	71
 	72 	def __repr__(self):

~/anaconda3/envs/my37/lib/python3.7/site-packages/xarray/coding/variables.py in _apply_mask(data, encoded_fill_values, decoded_fill_value, dtype)
	135 ) -> np.ndarray:
	136 	"""Mask all matching values in a NumPy arrays."""
--> 137 	data = np.asarray(data, dtype=dtype)
	138 	condition = False
	139 	for fv in encoded_fill_values:

~/anaconda3/envs/my37/lib/python3.7/site-packages/numpy/core/_asarray.py in asarray(a, dtype, order)
 	81
 	82 	"""
---> 83 	return array(a, dtype, copy=False, order=order)
 	84
 	85

~/anaconda3/envs/my37/lib/python3.7/site-packages/xarray/core/indexing.py in __array__(self, dtype)
	546
	547 	def __array__(self, dtype=None):
--> 548     	self._ensure_cached()
	549     	return np.asarray(self.array, dtype=dtype)
	550

~/anaconda3/envs/my37/lib/python3.7/site-packages/xarray/core/indexing.py in _ensure_cached(self)
	543 	def _ensure_cached(self):
	544     	if not isinstance(self.array, NumpyIndexingAdapter):
--> 545         	self.array = NumpyIndexingAdapter(np.asarray(self.array))
	546
	547 	def __array__(self, dtype=None):

~/anaconda3/envs/my37/lib/python3.7/site-packages/numpy/core/_asarray.py in asarray(a, dtype, order)
 	81
 	82 	"""
---> 83 	return array(a, dtype, copy=False, order=order)
 	84
 	85

~/anaconda3/envs/my37/lib/python3.7/site-packages/xarray/core/indexing.py in __array__(self, dtype)
	516
	517 	def __array__(self, dtype=None):
--> 518     	return np.asarray(self.array, dtype=dtype)
	519
	520 	def __getitem__(self, key):

~/anaconda3/envs/my37/lib/python3.7/site-packages/numpy/core/_asarray.py in asarray(a, dtype, order)
 	81
 	82 	"""
---> 83 	return array(a, dtype, copy=False, order=order)
 	84
 	85

~/anaconda3/envs/my37/lib/python3.7/site-packages/xarray/core/indexing.py in __array__(self, dtype)
	417 	def __array__(self, dtype=None):
	418     	array = as_indexable(self.array)
--> 419     	return np.asarray(array[self.key], dtype=None)
	420
	421 	def transpose(self, order):

~/anaconda3/envs/my37/lib/python3.7/site-packages/xarray/backends/netCDF4_.py in __getitem__(self, key)
 	90 	def __getitem__(self, key):
 	91     	return indexing.explicit_indexing_adapter(
---> 92         	key, self.shape, indexing.IndexingSupport.OUTER, self._getitem
 	93     	)
 	94

~/anaconda3/envs/my37/lib/python3.7/site-packages/xarray/core/indexing.py in explicit_indexing_adapter(key, shape, indexing_support, raw_indexing_method)
	708 	"""
	709 	raw_key, numpy_indices = decompose_indexer(key, shape, indexing_support)
--> 710 	result = raw_indexing_method(raw_key.tuple)
	711 	if numpy_indices.tuple:
	712     	# index the loaded np.ndarray

~/anaconda3/envs/my37/lib/python3.7/site-packages/xarray/backends/netCDF4_.py in _getitem(self, key)
	102         	with self.datastore.lock:
	103             	original_array = self.get_array(needs_lock=False)
--> 104             	array = getitem(original_array, key)
	105     	except IndexError:
	106         	# Catch IndexError in netCDF4 and return a more informative

~/anaconda3/envs/my37/lib/python3.7/site-packages/xarray/backends/common.py in robust_getitem(array, key, catch, max_retries, initial_delay)
 	63 	for n in range(max_retries + 1):
 	64     	try:
---> 65         	return array[key]
 	66     	except catch:
 	67         	if n == max_retries:

netCDF4/_netCDF4.pyx in netCDF4._netCDF4.Variable.__getitem__()

netCDF4/_netCDF4.pyx in netCDF4._netCDF4.Variable._get()

netCDF4/_netCDF4.pyx in netCDF4._netCDF4._ensure_nc_success()

RuntimeError: NetCDF: Access failure

**Comment by astrojuanlu** [comment]
We had the exact same issue (I'm using an old version of xarray, netCDF4, and Hyrax because I'm running an old app). I'm not sure when things broke, but what we ended up doing was configuring Tomcat to allow square brackets as advised in the mailing list thread posted by OP. We unfortunately don't have much time to debug this issue or add more info, but I hope this is useful for folks arriving here from search engines (as I did).
